recipe/thirdparty/providers/google.go

```diff
@@ -17,102 +17,179 @@ package providers
-		ID: googleID,
-		Get: func(redirectURI, authCodeFromRequest *string, userContext supertokens.UserContext) tpmodels.TypeProviderGetResponse {
-			accessTokenAPIURL := "https://accounts.google.com/o/oauth2/token"
-			accessTokenAPIParams := map[string]string{
-				"client_id":     config.ClientID,
-				"client_secret": config.ClientSecret,
-				"grant_type":    "authorization_code",
-			}
-			if authCodeFromRequest != nil {
-				accessTokenAPIParams["code"] = *authCodeFromRequest
-			}
-			if redirectURI != nil {
-				accessTokenAPIParams["redirect_uri"] = *redirectURI
-			}
+func Google(input tpmodels.TypeGoogleInput) tpmodels.TypeProvider {
+	googleProvider := &tpmodels.GoogleProvider{}
 
-			authorisationRedirectURL := "https://accounts.google.com/o/oauth2/v2/auth"
-			scopes := []string{"https://www.googleapis.com/auth/userinfo.email"}
-			if config.Scope != nil {
-				scopes = config.Scope
-			}
+	getConfig := func(clientID *string, userContext supertokens.UserContext) (tpmodels.GoogleConfig, error) {
+		if input.Config == nil || len(input.Config) == 0 {
+			return tpmodels.GoogleConfig{}, errors.New("please specify a config or override GetConfig")
+		}
 
-			var additionalParams map[string]interface{} = nil
-			if config.AuthorisationRedirect != nil && config.AuthorisationRedirect.Params != nil {
-				additionalParams = config.AuthorisationRedirect.Params
-			}
+		if clientID == nil && len(input.Config) > 1 {
+			return tpmodels.GoogleConfig{}, errors.New("please specify a clientID as there are multiple configs")
+		}
 
-			authorizationRedirectParams := map[string]interface{}{
-				"scope":                  strings.Join(scopes, " "),
-				"access_type":            "offline",
-				"include_granted_scopes": "true",
-				"response_type":          "code",
-				"client_id":              config.ClientID,
-			}
-			for key, value := range additionalParams {
-				authorizationRedirectParams[key] = value
+		if clientID == nil && len(input.Config) == 1 {
+			return input.Config[0], nil
+		}
+
+		for _, config := range input.Config {
+			if config.ClientID == *clientID {
+				return config, nil
 			}
+		}
+
+		return tpmodels.GoogleConfig{}, errors.New("config for specified clientID not found")
+	}
+
+	getAuthorisationRedirectURL := func(clientID *string, redirectURIOnProviderDashboard string, userContext supertokens.UserContext) (tpmodels.TypeAuthorisationRedirect, error) {
+		scopes := []string{"https://www.googleapis.com/auth/userinfo.email"}
+		config, err := (googleProvider.GetConfig)(clientID, userContext)
+		if err != nil {
+			return tpmodels.TypeAuthorisationRedirect{}, err
+		}
+		if config.Scope != nil {
+			scopes = config.Scope
+		}
+
+		queryParams := map[string]interface{}{
+			"scope":                  strings.Join(scopes, " "),
+			"access_type":            "offline",
+			"include_granted_scopes": "true",
+			"response_type":          "code",
+			"client_id":              getActualClientIdFromDevelopmentClientId(config.ClientID),
```

Comment:
Can the logic for development keys related stuff be moved to the main API logic?

Reply:
This can be done only after having clientID from the GetConfig. ClientID is optional to the APIs.